validator/client/validator.go

```diff
@@ -876,9 +878,23 @@ func (v *validator) domainData(ctx context.Context, epoch primitives.Epoch, doma
 	key := strings.Join([]string{strconv.FormatUint(uint64(req.Epoch), 10), hex.EncodeToString(req.Domain)}, ",")
 
 	if val, ok := v.domainDataCache.Get(key); ok {
+		v.domainDataLock.RUnlock()
 		return proto.Clone(val.(proto.Message)).(*ethpb.DomainResponse), nil
 	}
+	v.domainDataLock.RUnlock()
 
+	// Lock as we are about to perform an expensive request to the beacon node.
+	v.domainDataLock.Lock()
+
+	// We check the cache again as in the event there are multiple inflight requests for
+	// the same domain data, the cache might have been filled while we were waiting
+	// to acquire the lock.
+	if val, ok := v.domainDataCache.Get(key); ok {
+		v.domainDataLock.Unlock()
+		return proto.Clone(val.(proto.Message)).(*ethpb.DomainResponse), nil
+	}
+
+	defer v.domainDataLock.Unlock()
```

Comment:
```suggestion
	v.domainDataLock.Lock()
	defer v.domainDataLock.Unlock()

	// We check the cache again as in the event there are multiple inflight requests for
	// the same domain data, the cache might have been filled while we were waiting
	// to acquire the lock.
	if val, ok := v.domainDataCache.Get(key); ok {
		return proto.Clone(val.(proto.Message)).(*ethpb.DomainResponse), nil
	}
```

I think it's functionally the same to put the defer unlock immediately after the lock